File: WordBreak/WordBreak.py
class Solution(object):
    def wordBreak(self, s, wordDict):
        """
        :type s: str
        :type wordDict: Set[str]
        :rtype: bool
        """
        dp = [False for _ in range(len(s) + 1)]
        dp[0] = True
        for i in range(1, len(s) + 1):
            for j in range(i - 1, -1, -1):
                if dp[j]:
                    x = s[j:i]
                    if x in wordDict:
                        dp[i] = True
                        break
        return dp[len(s)]

File: WordBreak/WordBreak.cpp
class Solution {
public:
    bool wordBreak(string s, unordered_set<string>& wordDict) {
        vector<bool> dp(s.size() + 1, false);
        dp[0] = true;
        for(int i = 1; i <= s.size(); i++) {
        	for(int j = i - 1; j >= 0; j--) {
        		if(dp[j]) {
        			string sub = s.substr(j, i - j);
        			if(wordDict.find(sub) != wordDict.end()) {
        				dp[i] = true;
        				break;
        			}
        		}
        	}
        }
        return dp[s.size()];
	}
};